ducktape/tests/runner.py

```diff
@@ -61,46 +68,98 @@ def __init__(self, cluster, session_context, session_logger, tests, port=Console
+
+            # this handler should be a noop if we're in a client process, so it's an error if the current pid
+            # is in self._client_procs
+            assert p.pid != os.getpid(), "Signal handler should not reach this point in a client subprocess."
+            if p.is_alive():
+                os.kill(p.pid, signal.SIGTERM)
+
     def who_am_i(self):
         """Human-readable name helpful for logging."""
         return self.__class__.__name__
 
+    @property
+    def _ready_to_trigger_more_tests(self):
+        """Should we pull another test from the scheduler?"""
+        return not self.stop_testing and \
+            len(self.active_tests) < self.max_parallel and \
+            len(self.scheduler) > 0 and \
+            self.scheduler.peek() is not None
+
+    @property
+    def _expect_client_requests(self):
+        return len(self.active_tests) > 0
+
     def run_all_tests(self):
         self.results.start_time = time.time()
         self._log(logging.INFO, "starting test run with session id %s..." % self.session_context.session_id)
-        self._log(logging.INFO, "running %d tests..." % len(self.staged_tests))
+        self._log(logging.INFO, "running %d tests..." % len(self.scheduler))
 
-        while len(self.staged_tests) > 0 or len(self.active_tests) > 0:
+        while self._ready_to_trigger_more_tests or self._expect_client_requests:
 
-            if len(self.active_tests) == 0:
-                next_test_context = self.staged_tests.pop()
+            while self._ready_to_trigger_more_tests:
+                next_test_context = self.scheduler.next()
                 self._preallocate_subcluster(next_test_context)
                 self._run_single_test(next_test_context)
-            try:
-                event = self.receiver.recv()
-                self._handle(event)
-            except Exception as e:
-                err_str = "Exception receiving message: %s: %s" % (str(type(e)), str(e))
-                err_str += "\n" + traceback.format_exc(limit=16)
-                self._log(logging.ERROR, err_str)
-                continue
-
-        for proc in self.proc_list:
+
+            if self._expect_client_requests:
+                try:
+                    event = self.receiver.recv()
+                    self._handle(event)
+                except Exception as e:
+                    err_str = "Exception receiving message: %s: %s" % (str(type(e)), str(e))
+                    err_str += "\n" + traceback.format_exc(limit=16)
+                    self._log(logging.ERROR, err_str)
+                    continue
+
+        for proc in self._client_procs.values():
             proc.join()
 
         return self.results
 
+    def expected_num_nodes(self, test_context):
+        """Helper method for deciding how many nodes we expect the given test to use."""
+        expected = test_context.expected_num_nodes
+        return len(self.cluster) if expected is None else expected
```

Comment:
Thinking about this a bit more -- would it make sense to have  separate setting for max cluster size? I'm wondering how we're going to manage transitioning to having the cluster size annotations. If we actually allocated the whole cluster then we basically can't turn on parallel tests until we've got everything annotated. If we use some other value which can be set via command line flag or env var, then we can set a cluster size that matches today's cluster size given serial execution but still get parallelism without wasting the entire cluster's resources. Thoughts?